File: interview_cake/stock_price.rb
def get_max_profit(stock_prices)

  # Calculate the max profit.


  return 0
end


















# tests

def run_tests
  desc = 'price goes up then down'
  actual = get_max_profit([1, 5, 3, 2])
  expected = 4
  assert_equal(actual, expected, desc)

  desc = 'price goes down then up'
  actual = get_max_profit([7, 2, 8, 9])
  expected = 7
  assert_equal(actual, expected, desc)

  desc = 'price goes up all day'
  actual = get_max_profit([1, 6, 7, 9])
  expected = 8
  assert_equal(actual, expected, desc)

  desc = 'price goes down all day'
  actual = get_max_profit([9, 7, 4, 1])
  expected = -2
  assert_equal(actual, expected, desc)

  desc = 'price stays the same all day'
  actual = get_max_profit([1, 1, 1, 1])
  expected = 0
  assert_equal(actual, expected, desc)

  desc = 'one price raises error'
  assert_raises(desc) {
    get_max_profit([1])
  }

  desc = 'empty array raises error'
  assert_raises(desc) {
    get_max_profit([])
  }
end

def assert_equal(a, b, desc)
  if a == b
    puts "#{desc} ... PASS"
  else
    puts "#{desc} ... FAIL: #{a} != #{b}"
  end
end

def assert_raises(desc)
  begin
    yield
    puts "#{desc} ... FAIL"
  rescue
    puts "#{desc} ... PASS"
  end
end

run_tests()

File: interview_cake/stock_price.js
function getMaxProfit(stockPrices) {
  if (stockPrices.length < 2) {
    throw "Need at least 2 stock prices";
  }
  let min = stockPrices[0];
  let bestProfit = -1 / 0.0;

  for (let i = 1, len = stockPrices.length; i < len; i++) {
    const price = stockPrices[i];
    const profit = price - min;
    if (profit > bestProfit) { bestProfit = profit; }
    if (price < min) { min = price; }
  }

  return bestProfit;
}


















// tests

let desc = 'price goes up then down';
let actual = getMaxProfit([1, 5, 3, 2]);
let expected = 4;
assertEqual(actual, expected, desc);

desc = 'price goes down then up';
actual = getMaxProfit([7, 2, 8, 9]);
expected = 7;
assertEqual(actual, expected, desc);

desc = 'price goes up all day';
actual = getMaxProfit([1, 6, 7, 9]);
expected = 8;
assertEqual(actual, expected, desc);

desc = 'price goes down all day';
actual = getMaxProfit([9, 7, 4, 1]);
expected = -2;
assertEqual(actual, expected, desc);

desc = 'price stays the same all day';
actual = getMaxProfit([1, 1, 1, 1]);
expected = 0;
assertEqual(actual, expected, desc);

desc = 'one price throws error';
const onePrice = () => (getMaxProfit([1]));
assertThrowsError(onePrice, desc);

desc = 'empty array throws error';
const emptyArray = () => (getMaxProfit([]));
assertThrowsError(emptyArray, desc);

function assertEqual(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`)
  }
}

function assertThrowsError(func, desc) {
  try {
    func();
    console.log(`${desc} ... FAIL`);
  } catch (e) {
    console.log(`${desc} ... PASS`);
  }
}
